File: 0052 - N-Queens II/solution.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        def place(i):
            if i == n:
                return 1
            res = 0
            for j in range(n):
                if not (col[j] or dia1[i+j] or dia2[i-j+n-1]):
                    col[j] = dia1[i+j] = dia2[i-j+n-1] = True
                    res += place(i+1)
                    col[j] = dia1[i+j] = dia2[i-j+n-1] = False
            return res
        col, dia1, dia2 = [False]*n, [False]*(n*2-1), [False]*(n*2-1)
        return place(0)

File: 0052 - N-Queens II/solution.cpp
class Solution {
private:
    vector<bool> col, dia1, dia2;
    int place(int i, int n){
        if(i == n) return 1;
        int res = 0;
        for(int j = 0; j < n; j++){
            if(col[j] || dia1[i+j] || dia2[i-j+n-1]) continue;
            col[j] = dia1[i+j] = dia2[i-j+n-1] = true;
            res += place(i+1, n);
            col[j] = dia1[i+j] = dia2[i-j+n-1] = false;
        }
        return res;
    }
public:
    int totalNQueens(int n) {
        col.resize(n), dia1.resize(n*2-1), dia2.resize(n*2-1);
        return place(0, n);
    }
};